src/main.rs

```diff
@@ -18,50 +20,70 @@ mod crates_cache;
+    help: bool,
+    authors: bool,
+    publishers: bool,
+    crates: bool,
+    update: bool,
+    cache_max_age: Duration,
+    metadata_args: Vec<String>,
+}
 
-    while let Some(arg) = args.next() {
-        match arg.to_str() {
-            None => bail_bad_arg(arg),
-            Some("supply-chain") => (), // first arg when run as `cargo supply-chain`
-            Some("authors") => return subcommands::authors(args),
-            Some("publishers") => return subcommands::publishers(args),
-            Some("crates") => return subcommands::crates(args),
-            Some("update") => return subcommands::update(args),
-            Some(arg) if arg.starts_with("--") => bail_unknown_option(arg),
-            Some(arg) if arg.starts_with('-') => bail_unknown_short_option(arg),
-            Some(arg) => bail_unknown_command(arg),
+fn main() {
+    match args_parser() {
+        Ok(args) => handle_args(args),
+        Err(e) => {
+            eprintln!("Error {:?}", e);
+            eprint_help();
         }
     }
-
-    // No command selected.
-    bail_no_command();
 }
 
-fn bail_unknown_option(arg: &str) -> ! {
-    eprintln!("Unknown option: {}", std::path::Path::new(&arg).display());
-    eprint_help();
-    std::process::exit(1);
+fn handle_args(args: Args) {
+    if args.help {
+        eprint_help();
+    } else if args.authors {
+        subcommands::authors(args.metadata_args)
+    } else if args.publishers {
+        subcommands::publishers(args.metadata_args, args.cache_max_age)
+    } else if args.crates {
+        subcommands::crates(args.metadata_args, args.cache_max_age)
+    } else if args.update {
+        subcommands::update(args.cache_max_age)
+    } else {
+        eprint_help();
+    }
 }
 
-fn bail_unknown_short_option(arg: &str) -> ! {
-    eprintln!("Unknown flag: {}", arg);
-    eprint_help();
-    std::process::exit(1);
+fn parse_max_age(text: &str) -> Result<Duration, humantime::DurationError> {
+    humantime::parse_duration(&text)
 }
 
-fn bail_unknown_command(arg: &str) -> ! {
-    eprintln!("Unknown command: {}", arg);
-    eprint_help();
-    std::process::exit(1);
+fn parse_metadata_args(text: &str) -> Result<Vec<String>, &'static str> {
+    Ok(text
+        .split(" ")
+        .map(|s| String::from(s))
+        .collect::<Vec<String>>())
 }
 
-fn bail_no_command() -> ! {
-    eprintln!("No command selected.");
-    eprint_help();
-    std::process::exit(1);
+fn args_parser() -> Result<Args, pico_args::Error> {
+    let default_cache_max_age = Duration::from_secs(48 * 3600);
+    let mut args = Arguments::from_env();
+    let args = Args {
+        help: args.contains(["-h", "--help"]),
+        authors: args.contains(["-a", "--authors"]),
+        publishers: args.contains(["-p", "--publishers"]),
+        crates: args.contains(["-c", "--crates"]),
+        update: args.contains(["-u", "--update"]),
+        metadata_args: args
+            .opt_value_from_fn("--metadata-args", parse_metadata_args)?
```

Comment:
This changes the way metadata arguments are passed, but the help text still lists metadata arguments as being passed after `--`, and I'd prefer to keep it that way because this avoids double-escaping. Consider:

`cargo supply-chain crates -- --manifest-path="/path/to/file" --locked`

vs

`cargo supply-chain crates --metadata-args="--manifest-path=\"/path/to/file\" locked"`

when made into a parameter. The current behavior can be kept by using the [`free`](https://docs.rs/pico-args/0.3.4/pico_args/struct.Arguments.html#method.free) function from pico-args.

Reply:
I tried using free, but it doesn't seem to accept any other parameters that start with --, then it said UnexpectedArg, it seemed to only accept non-dashed arguments.

Reply:
I get this when trying the free version
```
/target/debug/cargo-supply-chain publishers -- --manifest-path="/path/to/file" --locked
Error UnusedArgsLeft(["--", "--manifest-path=/path/to/file", "--locked"])
```

Reply:
Okay, I got a solution for this one, I'll push when we resolved the rest of it.